cli/flox/src/commands/mod.rs

```diff
@@ -674,22 +690,136 @@ impl ConcreteEnvironment {
+/// An environment that has been activated with `flox activate`
+/// and can be (re)opened, i.e. to install packages into it.
+///
+/// Unlike [ConcreteEnvironment], this type does not hold concrete instances of environments,
+/// but rather fully qualified metadata to create an instance from.
+///
+/// * for [PathEnvironment] and [ManagedEnvironment] that's the path to their `.flox` and `.flox/pointer.json`
+/// * for [RemoteEnvironment] that's the [EnvironmentRef] to the remote environment
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub enum ActiveEnvironment {
+    /// Container for "local" environments pointed to by [DotFlox]
+    DotFlox(UninitializedEnvironment),
+    /// Container for [RemoteEnvironment]
+    Remote(EnvironmentRef),
+}
+
+impl ActiveEnvironment {
+    fn from_concrete_environment(env: &ConcreteEnvironment) -> Result<Self> {
+        match env {
+            ConcreteEnvironment::Path(path_env) => {
+                let pointer = path_env.pointer.clone().into();
+                Ok(Self::DotFlox(UninitializedEnvironment {
+                    path: path_env.parent_path().unwrap(),
+                    pointer,
+                }))
+            },
+            ConcreteEnvironment::Managed(managed_env) => {
+                let pointer = managed_env.pointer().clone().into();
+                Ok(Self::DotFlox(UninitializedEnvironment {
+                    path: managed_env.parent_path().unwrap(),
+                    pointer,
+                }))
+            },
+            ConcreteEnvironment::Remote(remote_env) => {
+                let env_ref = remote_env.env_ref();
+                Ok(Self::Remote(env_ref))
+            },
+        }
+    }
+}
+
+/// A list of environments that are currently active
+/// (i.e. have been activated with `flox activate`)
+///
+/// When inside a `flox activate` shell,
+/// flox stores [ActiveEnvironment] metadata to (re)open the activated environment
+/// in `$FLOX_ACTIVE_ENVIRONMENTS`.
 ///
-/// When inside a `flox activate` shell, flox stores the path to the
-/// activated environment in `$FLOX_ACTIVE_ENVIRONMENTS`. Environments which
-/// are activated while in a `flox activate` shell, are prepended - the most
-/// recently activated environment is the _first in the list of environments.
-/// TODO: how will we handle remote environments?
-fn last_activated_environment() -> Option<PathBuf> {
-    activated_environments().into_iter().next()
-}
-
-/// Return paths to all activated environments.
-fn activated_environments() -> Vec<PathBuf> {
-    env::var(FLOX_ACTIVE_ENVIRONMENTS_VAR)
-        .map(|active| env::split_paths(&active).collect())
-        .unwrap_or(vec![])
+/// Environments which are activated while in a `flox activate` shell, are prepended
+/// -> the most recently activated environment is the _first_ in the list of environments.
+///
+/// Internally this is implemented through a [VecDeque] which is serialized to JSON.
```

Comment:
question blocking: code wise it's not that complicated, but as a reader of this code, I think it's complicated/surprising that we're serializing a Rust structure as JSON which we're storing in an environment variable. Definitely cool and maybe even a bit more reliable, but I'm unsure if that's better than the simple, stupid approach of joining strings with `:`, which I think most people would expect. And it's a little odd to see `FLOX_ACTIVE_ENVIRONMENTS=[{"DotFlox":{"path":"/Users/matthew/flox/main","pointer":{"name":"main","version":1}}}]` in my environment. I'd want to hear more of why you decided to do it this way or get a sanity check from @zmitchell or @aakropotkin before merging.

Reply:
For DotFlox envs this was certainly _good enough_, its literally just alist of paths there.
With remote envs in the mix we need _some way_ to distinguish whether an env is a remote or a dot flox. One way was to check whether the path entry was absolute or relative. i.e. `/Users/me/home/project/ -> DotFloxEnv` and `me/project -> RemoteEnv`.
For one that appears rather hacky, for two we will likely need to also encode the floxhub url in there if we want to support multiple floxhubs _at once_ (thinking of, I should prob add that already..) urls will mess up the path separation since they contain `:`.
For json in your env i don't feel strongly opposed. If at all it gives you more flexibility/metadata to construct the `PS1`s.


--------

Does it look like this now?
> `{"DotFlox":{"path":"/Users/matthew/flox/main","pointer":{"name":"main","version":1}}}`

I guess that should be 
`{"type": "DotFlox", "path":"/Users/matthew/flox/main","pointer":{"name":"main","version":1}}`

but that's a quick fix

Reply:
I think it's fine to store data as JSON in an env var as long as the variable never exceeds ~130,000 characters.

If you're just storing simple key values something like `foo=bar:baz=quux:...` would be a bit more conventional; but if you have multiple fields for these objects JSON is likely the right choice.

Reply:
I'm seeing `FLOX_ACTIVE_ENVIRONMENTS=[{"type":"dot-flox","path":"/Users/matthew/flox/main","pointer":{"name":"main","version":1}}]` now